Mathlib/RingTheory/DualNumber.lean

```diff
@@ -0,0 +1,166 @@
+  simp only [isNilpotent_iff_eq_zero, Submodule.zero_eq_bot]
+  intros
+  exact Ideal.eq_bot_of_prime _
+
+end TrivSqZeroExt
+
+namespace DualNumber
+
+lemma fst_eq_zero_iff_eps_dvd {R : Type*} [Semiring R] {x : R[ε]} :
+    x.fst = 0 ↔ ε ∣ x := by
+  simp_rw [dvd_def, TrivSqZeroExt.ext_iff, TrivSqZeroExt.fst_mul, TrivSqZeroExt.snd_mul,
+    fst_eps, snd_eps, zero_mul, zero_smul, zero_add, MulOpposite.smul_eq_mul_unop,
+    MulOpposite.unop_op, one_mul, exists_and_left, iff_self_and]
+  intro
+  exact ⟨.inl x.snd, rfl⟩
+
+lemma isNilpotent_eps {R : Type*} [Semiring R] :
+    IsNilpotent (ε : R[ε]) :=
+  TrivSqZeroExt.isNilpotent_inr 1
+
+section Field
+
+open TrivSqZeroExt
+
+variable {K : Type*} [Field K]
+
+instance : LocalRing K[ε] :=
+  .of_isUnit_or_isUnit_one_sub_self fun _ ↦
+    (isUnit_or_isNilpotent _).imp_right IsNilpotent.isUnit_one_sub
+
+lemma isNilpotent_iff_eps_dvd {x : K[ε]} :
+    IsNilpotent x ↔ ε ∣ x := by
+  simp only [isNilpotent_iff_isNilpotent_fst, isNilpotent_iff_eq_zero, fst_eq_zero_iff_eps_dvd]
+
+lemma isMaximal_span_singleton_eps :
+    (Ideal.span {ε} : Ideal K[ε]).IsMaximal := by
+  rw [Ideal.isMaximal_iff]
+  simp only [Ideal.mem_span_singleton, ← isNilpotent_iff_eps_dvd, isNilpotent_iff_isNilpotent_fst,
+    fst_one, isNilpotent_iff_eq_zero, one_ne_zero, not_false_eq_true, true_and]
+  intro I x _ IH hx
+  rw [← Ideal.eq_top_iff_one]
+  rcases isUnit_or_isNilpotent x with hx'|hx'
+  · exact Ideal.eq_top_of_isUnit_mem _ hx hx'
+  · simp only [isNilpotent_iff_isNilpotent_fst, isNilpotent_iff_eq_zero] at hx'
+    exact absurd hx' IH
+
+lemma maximalIdeal_eq_span_singleton_eps :
+    LocalRing.maximalIdeal K[ε] = Ideal.span {ε} :=
+  (LocalRing.eq_maximalIdeal isMaximal_span_singleton_eps).symm
+
+lemma isMaximal_iff_span_singleton_eps {I : Ideal K[ε]} :
+    I.IsMaximal ↔ I = .span {ε} := by
+  rw [← maximalIdeal_eq_span_singleton_eps]
+  constructor
+  · exact LocalRing.eq_maximalIdeal
+  · rintro rfl
+    infer_instance
+
+instance : IsPrincipalIdealRing K[ε] where
+  principal I := by
+    rcases eq_or_ne I ⊥ with rfl|hb
+    · exact bot_isPrincipal
+    rcases eq_or_ne I ⊤ with rfl|ht
+    · exact top_isPrincipal
+    obtain ⟨x, hxI, hx0⟩ := Submodule.exists_mem_ne_zero_of_ne_bot hb
+    refine ⟨x, le_antisymm ?_ ((Ideal.span_singleton_le_iff_mem I).mpr hxI)⟩
+    rcases isUnit_or_isNilpotent x with hx|hx
+    · simp [Ideal.eq_top_of_isUnit_mem _ hxI hx] at ht
+    simp only [isNilpotent_iff_isNilpotent_fst, isNilpotent_iff_eq_zero] at hx
+    rw [← inl_fst_add_inr_snd_eq x, hx, inl_zero, zero_add, inr_eq_smul_eps,
+      Ideal.submodule_span_eq, ← inl_mul_eq_smul, Ideal.span_singleton_mul_left_unit,
+      ← maximalIdeal_eq_span_singleton_eps]
+    · exact LocalRing.le_maximalIdeal ht
+    · contrapose! hx0
+      simp only [isUnit_inl_iff, isUnit_iff_ne_zero, ne_eq, not_not] at hx0
+      ext <;> simp [hx, hx0]
+
+lemma exists_mul_left_or_mul_right (a b : K[ε]) :
+    ∃ c, a * c = b ∨ b * c = a := by
```

Comment:
I think this is true very generally for semirings if the same is true for the `.fst` projections; the lemmas already exist in `TrivSqZeroExt.lean` which construct the explicit left and right inverses.

Reply:
Ah, I meant something like
```
lemma exists_mul_left_or_mul_right (a b : tsze R M) (h : ∃ c, a.fst * c.fst = b.fst ∨ b.fst * c.fst = a.fst) :
    ∃ c, a * c = b ∨ b * c = a := by
```
which I think follows from `TrivSqZeroExt.mul_right_eq_one` and similar

Reply:
Maybe the `h : ` generalization is pointless, but you should be able to generalize to `tsze R M`.

Reply:
I still don't see how to prove it for `tsze` general over `M` (with plain AddCommGroup M), since one needs to somehow cancel out `b.snd` to get `a.snd` here:
```
theorem TrivSqZeroExt.extracted_1.{v, u} {R : Type u} {M : Type v} [inst : DivisionSemiring R] [inst_1 : AddCommGroup M]
  [inst_2 : Module Rᵐᵒᵖ M] [inst_3 : Module R M] [inst_4 : SMulCommClass R Rᵐᵒᵖ M] (a b : tsze R M) (ha : a.fst = 0)
  (hb : b.fst = 0) (hb' : b.snd ≠ 0) : ∃ c, a * c = b ∨ b * c = a := sorry
```

Reply:
I still don't see how to prove it for `tsze` general over `M` (with plain AddCommGroup M), since one needs to somehow cancel out `b.snd` to get `a.snd` here:
```
theorem TrivSqZeroExt.extracted_1.{v, u} {R : Type u} {M : Type v} [inst : DivisionSemiring R] [inst_1 : AddCommGroup M]
  [inst_2 : Module Rᵐᵒᵖ M] [inst_3 : Module R M] [inst_4 : SMulCommClass R Rᵐᵒᵖ M] (a b : tsze R M) (ha : a.fst = 0)
  (hb : b.fst = 0) (hb' : b.snd ≠ 0) : ∃ c, a * c = b ∨ b * c = a := sorry
```

Reply:
The goal simplifies to `∃ c, a.snd <• c.fst = b.snd ∨ b.snd <• c.fst = a.snd` but there is no way to stuff the `a.snd : M` into a `c.fst`, the types don't match.

Reply:
Sorry for pointing you down a dead end!